viewing_party/party.py

```diff
@@ -1,23 +1,297 @@
+    # (do not prematurely optimize, do not prematurely optimize, ...)
+
+    friends_movie_list = []
+    for friend in user_data['friends']:
+        friends_movie_list.extend(friend['watched'])
+    
+    return friends_movie_list 
+
+def get_unique_watched(user_data):
+    # takes user_data returns a list of movies watched by 
+    # the user but not watched by friends
+    
+    # all of these should be lists 
+    friends_watched = watched_by_friends(user_data)
+    my_watched = user_data.get('watched', [])
+    my_unique = []
+
+    for movie in my_watched:
+        if movie not in friends_watched:
+            my_unique.append(movie)
+
+    return my_unique
+
+def remove_duplicates_by_title(movie_list):
+    # remove duplicates from a list using the title as a key
+    # not ideal to use titles alone as keys due to duplicate 
+    # titles, but it gets the job done. 
+
+    movie_dict = {}
+    for movie in movie_list:
+        if movie['title'] not in movie_dict:
+            movie_dict[movie['title']] = movie 
+    
+    # make sure the fuction returns a list rather than 
+    # a dict.values() iterator
+    return list(movie_dict.values())
+
+
+def get_friends_unique_watched(user_data):
+    # takes user_data returns a list of movies watched by 
+    # friends but not watched by the user 
+
+    # all of these should be lists 
+    friends_watched = watched_by_friends(user_data)
+    my_watched = user_data.get('watched', [])
+    friends_not_unique = []
+
+    for movie in friends_watched:
+        if movie not in my_watched:
+            friends_not_unique.append(movie)
+
+    # remove duplicates from friends_not_unique
+    friends_unique = remove_duplicates_by_title(friends_not_unique)
+
+    return friends_unique
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    # return a list of recommendations from friends
+    # that are available through subscribed streaming services
+
+    subscriptions = user_data['subscriptions']
+    movies_from_friends = remove_duplicates_by_title(watched_by_friends(user_data))
+    my_watched_titles = [movie['title'] for movie in user_data['watched']]
+
+    recs = []
+    for movie in movies_from_friends:
+        if movie['title'] not in my_watched_titles:
+            if movie['host'] in subscriptions:
+                recs.append(movie)
+    return recs 
+        
+            
+
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def most_frequent_genre(user_data):
+    # get the most frequent genre from user_data
+
+    genre_freq_table = {}
+    for movie in user_data['watched']:
+        if movie['genre'] in genre_freq_table:
+            genre_freq_table[movie['genre']] += 1
+        else:
+            genre_freq_table[movie['genre']] = 1

```

Comment:
The main difference between this version and your previous version is the other one used `get`, which I prefer over this method.